src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py

```diff
@@ -58,13 +75,91 @@ def _validation(target, data, evaluators, output_path, tracking_uri, evaluation_
+    :rtype: List[str]
+    :raises: ValueError
+    """
     try:
         data_df = pd.read_json(data, lines=True)
     except Exception as e:
-        raise ValueError(f"Failed to load data from {data}. Please validate it is a valid jsonl data. Error: {str(e)}.")
+        raise ValueError(
+            f"Failed to load data from {data}. Please validate it is a valid jsonl data. Error: {str(e)}.")
+    return list(data_df.columns)
 
-    for evaluator_name, evaluator in evaluators.items():
-        _validate_input_data_for_evaluator(evaluator, evaluator_name, data_df)
+
+def _validate_columns(columns: List[str], evaluators: Dict[str, Any], target: Optional[Callable]) -> None:
+    """
+    Check that all columns needed by evaluator or target function are present.
+
+    :keyword columns: The list of column in the data frame.
+    :paramtype columns: List[str]
+    :keyword evaluators: The dictionary of evaluators.
+    :paramtype evaluators: Dict[str, Any]
+    :keyword target: The callable to be applied to data set.
+    :paramtype target: Optional[Callable]
+    """
+    if target:
+        # If the target function is given, it may return
+        # several columns and hence we cannot check the availability of columns
+        # without knowing target function semantics.
+        # Instead, here we will validate the columns, taken by target.
+        _validate_input_data_for_evaluator(target, None, columns, is_target_fn=True)
+    else:
+        for evaluator_name, evaluator in evaluators.items():
+            _validate_input_data_for_evaluator(evaluator, evaluator_name, columns)
+
+
+def _apply_target_to_data(target: Callable, data: str, pf_client: PFClient) -> str:
+    """
+    Apply the target function to the data set and save the data to temporary file.
+
+    :keyword target: The function to be applied to data.
+    :paramtype target: Callable
+    :keyword data: The path to input jsonl file.
+    :paramtype data: str
+    :keyword pf_client: The promptflow client to be used.
+    :paramtype pf_client: PFClient
+    :return: The path to data file with answers from target function.
+    :rtype: str
+    """
+    # We are manually creating the temporary directory for the flow
+    # because the way tempdir remove temporary directories will
+    # hang the debugger, because promptflow will keep flow directory.
+    saved_flow = f'flow_{uuid.uuid1()}'
+    os.makedirs(saved_flow)
+    save_function_as_flow(fun=target, target_dir=saved_flow, pf=pf_client)
+    run = pf_client.run(
+        flow=saved_flow,
+        data=data,
+        name=f'preprocess_{uuid.uuid1()}'
+    )
+    run = pf_client.stream(run)
+    # Delete temporary directory if we can.
+    try:
+        shutil.rmtree(saved_flow)
+    except BaseException:
+        # Exception means, we are running in debugger. In this case we can keep the
+        # directory.
+        pass
+    function_output = pd.read_json(pf_client.runs._get_outputs_path(run),

```

Comment:
There is a pf api to get output as dataframe. Should we use it directly ? I think its called `get_results`

Reply:
Fixed